File: arrays/02_max_monster_eliminated/max_monster_eliminated.py
from typing import List


class Solution:
    def max_monster_eliminated(self, dist: List[int], speed: List[int]) -> int:
        if not self._check_constraints(dist, speed):
            raise ValueError("Constraints violated")

        n = len(dist)
        # Calculate time to reach the city for each monster
        time_to_city = [dist[i] / speed[i] for i in range(n)]
        time_to_city.sort()

        monsters_eliminated = 0
        for i in range(n):
            if time_to_city[i] <= i:
                return monsters_eliminated

            monsters_eliminated += 1

        return n

    def _check_constraints(self, dist: List[int], speed: List[int]) -> bool:
        n = len(dist)
        m = len(speed)
        if n != m:
            return False

        if n < 1 or n > 10**5:
            return False

        for i in range(n):
            if dist[i] < 1 or dist[i] > 10**5:
                return False
            if speed[i] < 1 or speed[i] > 10**5:
                return False

        return True

File: arrays/02_max_monster_eliminated/max_monster_eliminated.cpp
#include <vector>
#include <cassert>
#include <algorithm>
#include <cmath>
#include <stdexcept>

using namespace std;


class Solution {
public:
    int maxMonsterEliminated(const vector<int>& dist, const vector<int>& speed) {
        if (!checkConstraints(dist, speed)) {
            throw std::runtime_error("Constraints violated");
        }
        int n = dist.size();
        // Calculate time to reach the city for each monster
        vector<double> timeToCity(n);
        for (int i = 0; i < n; ++i) {
            timeToCity[i] = static_cast<double>(dist[i]) / speed[i];
        }
        // Sort the times to eliminate those monsters posing the immediate threats
        sort(timeToCity.begin(), timeToCity.end());

        int monstersEliminated = 0;
        for (int i = 0; i < n; ++i) {
            if (timeToCity[i] <= i) {
                return monstersEliminated;
            }
            monstersEliminated++;
        }

        return n;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const vector<int>& dist, const vector<int>& speed) {
        int n = dist.size(); 
        int m = speed.size();
        if (m != n) return false;

        if (n < 1 || n > pow(10, 5)) return false;

        for (int i = 0; i < n; ++i) {
            if (dist[i] < 1 || dist[i] > pow(10, 5)) return false;
            if (speed[i] < 1 || speed[i] > pow(10, 5)) return false;
        }

        return true;
    }
};
